### PR TITLE
Refactor app/pkg init logic

### LINKED ISSUE #827: `kctrl`: Post-alpha refactoring tracker for package authoring commands
Tasks at hand:
- [x] Refactor vendor utils to be a scoped struct with associated functions
- [x] Refactor workflows to move away from the step interface and update them according to refactoring described above
    - [x] Local directory
    - [x] Generic vendir step
    - [x] Git release workflow
    - [x] Files from git repo
    - [x] Helm chart from git repo
    - [x] Helm chart from OCI repository
- [x] Dedup and break up `package init` logic into smaller pieces
- [x] Move away from file utils and use existing art in in-built go packages (`ioutil`, `os`) instead
- [x] Refactor existing structs into packages
- [x] Handle how fetch option is passed between stages better
- [x] Verify that there is no regressions through means of existing tests and a bit of manual testing

### PR DESCRIPTION
**Pull request by 100mik**
#### What this PR does / why we need it:
Refactors and cleans up app/package init logic

#### Which issue(s) this PR fixes:

Fixes #827 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

##### Review Checklist:

- [x] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [x] Relevant tests are added or updated
- [x] Relevant docs in this repo added or updated
- [x] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [x] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by 100mik** [comment]
Gotta work towards wrapping up pending cleanup on this through a bunch of reviews @praveenrewar

File: cli/pkg/kctrl/cmd/package/init/pkg_build.go [hunk 1]
```diff
@@ -73,11 +77,61 @@ func (b *PackageBuild) SetObjectMeta(metaObj *metav1.ObjectMeta) {
...
+		if appTemplate.HelmTemplate != nil {
+			return true
+		}
+	}
+	return false
+}
+
+func (b *PackageBuild) ConfigureExportSection() {
+	fetchSource := b.GetObjectMeta().Annotations[appbuild.FetchContentAnnotationKey]
+	exportSection := *b.GetExport()
```
Comment (rohitagg2020) [inline cli/pkg/kctrl/cmd/package/init/pkg_build.go]:
```suggestion
	exportSection := b.GetExport()
```

File: cli/pkg/kctrl/cmd/app/init/git.go [hunk 2]
```diff
@@ -11,62 +11,55 @@ import (
...
 
-func (g *GitStep) PreInteract() error { return nil }
-
-func (g *GitStep) Interact() error {
-	contents := g.vendirConfig.Directories[0].Contents
+func (g *GitConfiguration) Configure() error {
+	contents := g.vendirConfig.Contents()
 	if contents == nil {
-		err := g.initializeContentWithGit()
+		err := g.initializeContentWithGit(contents)
```
Comment (rohitagg2020) [inline cli/pkg/kctrl/cmd/app/init/git.go]: As `g` already has the information about `contents`, IMO we can skip passing contents.

File: cli/pkg/kctrl/cmd/app/init/git.go [hunk 3]
```diff
@@ -11,62 +11,55 @@ import (
...
 	}
-	return g.getIncludedPaths()
+	return g.getIncludedPaths(contents)
 }
 
-func (g *GitStep) PostInteract() error { return nil }
-
-func (g *GitStep) initializeContentWithGit() error {
+func (g *GitConfiguration) initializeContentWithGit(contents []vendirconf.DirectoryContents) error {
 	//TODO Rohit need to check this how it should be done. It is giving path as empty.
```
Comment (rohitagg2020) [inline cli/pkg/kctrl/cmd/app/init/git.go]: This comment can be removed.

File: cli/pkg/kctrl/local/sources/fetch.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+const (
+	FetchFromGithubRelease       string = "Github Release"
+	FetchFromGit                 string = "Git Repository"
+	FetchFromHelmRepo            string = "Helm Chart from Helm Repository"
+	FetchChartFromGit            string = "Helm Chart from Git Repository"
+	FetchFromLocalDirectory      string = "Local Directory"
+	MultipleFetchOptionsSelected string = "MultipleFetchOptionsSelected"
+)
+
+type FetchConfiguration struct {
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/fetch.go]: `FetchConfiguration` -> `Source`? (How does `Source.Configure` sound compared to `FetchConfiguration.Configure`?)

File: cli/pkg/kctrl/local/sources/fetch.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	vendirFileName      = "vendir.yml"
+	includeAllFiles     = "*"
+)
+
+const (
+	FetchFromGithubRelease       string = "Github Release"
+	FetchFromGit                 string = "Git Repository"
+	FetchFromHelmRepo            string = "Helm Chart from Helm Repository"
+	FetchChartFromGit            string = "Helm Chart from Git Repository"
+	FetchFromLocalDirectory      string = "Local Directory"
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/fetch.go]: I think we can remove `FetchFrom` as the package name itself is source? Also, should we create a type for these?

File: cli/pkg/kctrl/local/sources/fetch.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	f.ui.PrintInformationalText("Please provide the location from where your Kubernetes manifests or Helm chart can be fetched. This will be bundled as a part of the package.")
+
+	vendirConfig := NewVendirConfig(vendirFileName)
+	err := vendirConfig.Load()
+	if err != nil {
+		return "", nil, err
+	}
+
+	isHelmTemplateExistInPreviousOption := f.build.HasHelmTemplate()
+	previousFetchOptionSelected := vendirConfig.FetchMode(isHelmTemplateExistInPreviousOption)
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/fetch.go]: inline?

File: cli/pkg/kctrl/local/sources/fetch.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+	if previousFetchOptionSelected == MultipleFetchOptionsSelected {
+		// As this is advanced use case, we dont know how to handle it.
+		f.ui.PrintInformationalText("Since vendir is syncing data from multiple resources, we will not reconfigure vendir.yml and run vendir sync.")
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/fetch.go]: Can we add a TODO instead of this comment?
Also, the informational text seems to be bit ambiguous, are we saying we will run `vendir sync` or are we saying we won't run `vendir sync`?

File: cli/pkg/kctrl/local/sources/fetch.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+		return "", nil, nil
+	}
+
+	options := []string{FetchFromLocalDirectory, FetchFromGithubRelease, FetchFromHelmRepo, FetchFromGit, FetchChartFromGit}
+	previousFetchOptionIndex := f.getPreviousFetchOptionIndex(options, previousFetchOptionSelected)
+	defaultFetchOptionIndex := previousFetchOptionIndex
+	choiceOpts := ui.ChoiceOpts{
+		Label:   "Enter source",
+		Default: defaultFetchOptionIndex,
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/fetch.go]:
```suggestion
		Default: f.getPreviousFetchOptionIndex(options, previousFetchOptionSelected),
```

File: cli/pkg/kctrl/local/sources/fetch.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+	}
+	currentFetchOptionIndex, err := f.ui.AskForChoice(choiceOpts)
+	if err != nil {
+		return "", nil, err
+	}
+	currentFetchOptionSelected := options[currentFetchOptionIndex]
+
+	if previousFetchOptionSelected != "" && currentFetchOptionSelected != previousFetchOptionSelected {
+		return "", nil, fmt.Errorf("Transitioning from one fetch option to another is not allowed. Earlier option selected: %s, Current Option selected: %s", previousFetchOptionSelected, currentFetchOptionSelected)
+	}
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/fetch.go]: If we don't allow transitioning from one source to another, maybe we shouldn't ask that question and just go ahead with the previous option along with a message?

Comment (praveenrewar) [reply cli/pkg/kctrl/local/sources/fetch.go]: If we consider working on this, maybe a separate PR would make sense.

File: cli/pkg/kctrl/local/sources/fetch.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+	vendirDirectories := vendirConfig.Directories()
+	if len(vendirDirectories) > 1 {
+		return currentFetchOptionSelected, nil, fmt.Errorf("More than 1 directory config found in the vendir file. (hint: Run vendir sync manually)")
+	}
+	if len(vendirDirectories) == 0 {
+		err := f.initializeVendirDirectorySection(vendirConfig)
+		if err != nil {
+			return currentFetchOptionSelected, nil, err
+		}
+	} else {
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/fetch.go]: `switch`?

File: cli/pkg/kctrl/local/sources/fetch.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+		previousFetchOptionIndex = 0
+	} else {
+		for i, fetchTypeName := range manifestOptions {
+			if fetchTypeName == previousFetchOption {
+				previousFetchOptionIndex = i
+				break
+			}
+		}
+	}
+	return previousFetchOptionIndex
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/fetch.go]:
```suggestion
	for i, fetchTypeName := range manifestOptions {
		if fetchTypeName == previousFetchOption {
			return i
		}
	}
	return 0
```

File: cli/pkg/kctrl/local/sources/fetch.go [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+func (f *FetchConfiguration) initializeVendirDirectorySection(vendirConfig *VendirConfig) error {
+	var directory vendirconf.Directory
+	directory = vendirconf.Directory{
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/fetch.go]: inline?

File: cli/pkg/kctrl/local/sources/fetch.go [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+		Path: VendirSyncDirectory,
+		Contents: []vendirconf.DirectoryContents{
+			{
+				Path: ".",
+			},
+		},
+	}
+	directories := []vendirconf.Directory{directory}
+	vendirConfig.SetDirectories(directories)
+	err := vendirConfig.Save()
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/fetch.go]: It seems like this whole function should be a method on vendirConfig instead?

File: cli/pkg/kctrl/local/sources/vendir_runner.go [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+func (r VendirRunner) RunSync() error {
+	_, err := os.Stat(vendirFileName)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if os.IsNotExist(err) {
+		return nil
+	}
+
+	r.ui.PrintInformationalText("We will use vendir to fetch the data from the source to the local directory. Vendir allows us to declaratively state what should be in a directory and sync data sources into it. All the information entered above has been persisted into a vendir.yml file.")
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/vendir_runner.go]: Maybe split it across multiple lines?

File: cli/pkg/kctrl/local/sources/vendir_runner.go [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+
+type VendirRunner struct {
+	ui cmdcore.AuthoringUI
+}
+
+func NewVendirRunner(ui cmdcore.AuthoringUI) VendirRunner {
+	return VendirRunner{ui: ui}
+}
+
+func (r VendirRunner) RunSync() error {
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/vendir_runner.go]: Since it's already called VendirRunner, maybe we should just use `Sync` instead of `RunSync`?

File: cli/pkg/kctrl/local/sources/vendir_runner.go [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+	_, err := os.Stat(vendirFileName)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if os.IsNotExist(err) {
+		return nil
+	}
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/vendir_runner.go]: I guess sometimes it's okay to have nested `if`s
```suggestion
	if err != nil {
	        if os.IsNotExist(err) {
		       return nil
	        }
		return err
	}
```
BTW, what other kind of errors are we expecting here?

File: cli/pkg/kctrl/local/sources/vendir_runner.go [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+		return err
+	}
+	err = r.runVendirSync()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r VendirRunner) printVendirFile() error {
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/vendir_runner.go]: I feel that this is not needed. We can directly use `PrintActionableText` and `printFile`

File: cli/pkg/kctrl/local/sources/vendir_runner.go [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+func (r VendirRunner) printFile(filePath string) error {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return fmt.Errorf("Printing file: %w", err)
+	}
+	defer func() {
+		file.Close()
+	}()
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/vendir_runner.go]: `defer file.Close()`?

File: cli/pkg/kctrl/local/sources/vendir_runner.go [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+	r.ui.PrintActionableText("Running vendir sync")
+	r.ui.PrintCmdExecutionText("vendir sync -f vendir.yml\n")
+	var stdoutBs, stderrBs bytes.Buffer
+
+	localCmdRunner := exec.NewPlainCmdRunner()
+	cmd := goexec.Command("vendir", []string{"sync", "-f", vendirFileName}...)
+	cmd.Stdin = nil
+	cmd.Stdout = &stdoutBs
+	cmd.Stderr = &stderrBs
+	err := localCmdRunner.Run(cmd)
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/vendir_runner.go]: We might have a good reason but I am not sure why we are not using `cmd.Run()`

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 4 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: Definitely better, 
I think the returned values make sense in this context as well

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 6 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: Yep makes sense

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 5 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: I actually did create a type `FetchMode`, with functions `IsLocal()` and `String()`
Since they are directly used as string options while prompting the user directly, it was making things a bit verbose

But I am happy to make the change

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 7 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: I will clean it up, we made a call to not handle this at all, so we can get rid of it

File: cli/pkg/kctrl/local/sources/vendir_runner.go (hunk 15 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/vendir_runner.go]: Makes sense!

File: cli/pkg/kctrl/local/sources/vendir_runner.go (hunk 16 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/vendir_runner.go]: Most I can think of is lack of permissions

File: cli/pkg/kctrl/local/sources/vendir_runner.go (hunk 18 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/vendir_runner.go]: Yep, definitely

File: cli/pkg/kctrl/local/sources/vendir_runner.go (hunk 19 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/vendir_runner.go]: I will need to refresh my memory here

File: cli/pkg/kctrl/cmd/package/init/pkg_build.go (hunk 1 above)
Comment (100mik) [reply cli/pkg/kctrl/cmd/package/init/pkg_build.go]: We won't be able to do length checks on `Export` if we make that change.
But I guess `GetExport` could return a non-pointer value

File: cli/pkg/kctrl/cmd/app/init/git.go (hunk 2 above)
Comment (100mik) [reply cli/pkg/kctrl/cmd/app/init/git.go]: Sticking to this because all of the subsequent calls use contents, and we would end up calling `Contents()` multiple times. Can switch it up if you feel stronglyabout this

File: cli/pkg/kctrl/cmd/package/init/pkg_build.go (hunk 1 above)
Comment (100mik) [reply cli/pkg/kctrl/cmd/package/init/pkg_build.go]: Made this change ^

File: cli/pkg/kctrl/cmd/app/init/git.go (hunk 3 above)
Comment (100mik) [reply cli/pkg/kctrl/cmd/app/init/git.go]: Done

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 4 above)
Comment (praveenrewar) [reply cli/pkg/kctrl/local/sources/fetch.go]: Although something to think about here is that the package name is sources, so outside this function it would be `sources.Source` 🤔

Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: Also renaming GitConfiguration to GitSource, and the others too along similar lines

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 5 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: Removed the prefix for now, leaving this open if we wanna have a type

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 4 above)
Comment (praveenrewar) [reply cli/pkg/kctrl/local/sources/fetch.go]: Yep, was about to add that as a comment 👍🏻

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 7 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: And to clarify we basically won't overwrite the file and run vendor sync, making it a bit clearer

Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: Done, let me know if it is satisfactory once I push the changes (leaving this open for your to hit resolve)

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 9 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: Yep, let's do a separate one for this
Closing this after I have created an issue

File: cli/pkg/kctrl/cmd/package/init.go [hunk 20]
```diff
@@ -64,54 +64,63 @@ func (o *InitOptions) Run() error {
...
 	}
-	pkgBuild.Save()
+
+	// Source does not need to be conifgured if manifests are in the local directory
+	if sourceConfiguration != nil {
+		err = sourceConfiguration.Configure()
+		if err != nil {
+			return err
+		}
+	}
```
Comment (praveenrewar) [inline cli/pkg/kctrl/cmd/package/init.go]: Should this be directly handled in the main `Configure` function?

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 8 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: Done!

File: cli/pkg/kctrl/cmd/package/init.go [hunk 21]
```diff
@@ -215,6 +224,22 @@ func (o *InitOptions) newOrExistingPackageResources() (*v1alpha1.Package,
 	return pkg, pkgMetadata, pkgInstall, nil
 }
 
+func (o *InitOptions) updateAndDefaultPackageResources(pkgRefName string, pkgMetadata *v1alpha1.PackageMetadata, pkg *v1alpha1.Package) {
+	pkgMetadata.Name = pkgRefName
+	pkg.Spec.RefName = pkgRefName
+	pkgMetadata.Spec.DisplayName = strings.Split(pkgRefName, ".")[0]
+
+	shortDesc := pkgMetadata.Spec.ShortDescription
+	if len(shortDesc) == 0 {
```
Comment (praveenrewar) [inline cli/pkg/kctrl/cmd/package/init.go]: inline? 
What if this was removed by the user?

File: cli/pkg/kctrl/cmd/package/init.go [hunk 22]
```diff
@@ -215,6 +224,22 @@ func (o *InitOptions) newOrExistingPackageResources() (*v1alpha1.Package,
...
+		pkgMetadata.Spec.ShortDescription = pkgRefName
+	}
+
+	longDesc := pkgMetadata.Spec.LongDescription
+	if len(longDesc) == 0 {
```
Comment (praveenrewar) [inline cli/pkg/kctrl/cmd/package/init.go]: Same as above.

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 10 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: Done!

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 11 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: Makes sense!

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 4 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: >Although something to think about here is that the package name is sources, so outside this function it would be sources.Source 🤔

I think that is fine (Example: `resources.Resource`)

File: cli/pkg/kctrl/local/sources/git.go [hunk 23]
```diff
@@ -11,62 +11,61 @@ import (
...
+
+func (g *GitConfiguration) Configure() error {
+	contents := g.vendirConfig.Contents()
 	if contents == nil {
-		err := g.initializeContentWithGit()
+		err := g.initializeContentWithGit(contents)
 		if err != nil {
 			return err
 		}
 	} else if contents[0].Git == nil {
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/git.go]: Should there be a check on len of contents?

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 12 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: Done!

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 13 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: Makes sense, making the move

File: cli/pkg/kctrl/local/sources/git.go [hunk 24]
```diff
@@ -11,62 +11,61 @@ import (
...
-	return g.getIncludedPaths()
+	return g.getIncludedPaths(contents)
 }
 
-func (g *GitStep) PostInteract() error { return nil }
-
-func (g *GitStep) initializeContentWithGit() error {
+func (g *GitConfiguration) initializeContentWithGit(contents []vendirconf.DirectoryContents) error {
 	//TODO Rohit need to check this how it should be done. It is giving path as empty.
-	g.vendirConfig.Directories[0].Contents = append(g.vendirConfig.Directories[0].Contents, vendirconf.DirectoryContents{})
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/git.go:57]: In the configure we have contents as `g.vendirConfig.Contents()` but here it is `g.vendirConfig.Directories[0].Contents`, is this change intentional?

File: cli/pkg/kctrl/local/sources/git.go [hunk 25]
```diff
@@ -11,62 +11,61 @@ import (
...
-	contents := g.vendirConfig.Directories[0].Contents
+func NewGitConfiguration(ui cmdcore.AuthoringUI, vendirConfig *VendirConfig) *GitConfiguration {
+	return &GitConfiguration{ui: ui, vendirConfig: vendirConfig}
+}
+
+func (g *GitConfiguration) Configure() error {
+	contents := g.vendirConfig.Contents()
 	if contents == nil {
-		err := g.initializeContentWithGit()
+		err := g.initializeContentWithGit(contents)
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/git.go]: I agree with @rohitagg2020, passing contents seems to be creating a lot of noise here.

File: cli/pkg/kctrl/local/sources/vendir_runner.go (hunk 14 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/vendir_runner.go]: Done

File: cli/pkg/kctrl/local/sources/vendir_runner.go (hunk 15 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/vendir_runner.go]: Done!

File: cli/pkg/kctrl/local/sources/vendir_runner.go (hunk 17 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/vendir_runner.go]: done!

File: cli/pkg/kctrl/cmd/package/init.go (hunk 20 above)
Comment (100mik) [reply cli/pkg/kctrl/cmd/package/init.go]: We will get a nil pointer exception 😢

Comment (100mik) [reply cli/pkg/kctrl/cmd/package/init.go]: Before it reaches the function

File: cli/pkg/kctrl/cmd/package/init.go (hunk 21 above)
Comment (100mik) [reply cli/pkg/kctrl/cmd/package/init.go]: Good question, but I cannot imagine how we will figure that out 🤔 
Since we are only setting doing this for values that are displayed by kctrl while listing/getting

I guess nudging users towards giving them values is not the worst idea 😆

Comment (100mik) [reply cli/pkg/kctrl/cmd/package/init.go]: I guess we can discuss this and put up an issue if we wanna give it more thought

File: cli/pkg/kctrl/cmd/package/init.go (hunk 22 above)
Comment (100mik) [reply cli/pkg/kctrl/cmd/package/init.go]: Needs more discussion as I mentioned above

File: cli/pkg/kctrl/local/sources/git.go (hunk 25 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/git.go]: We will be replacing `contents[0]` with `vendirConfig.Contents()[0]` in all the functions if we want to do that, true for all of these cases
Happy to do it if y'all think I should

File: cli/pkg/kctrl/local/sources/git.go (hunk 23 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/git.go]: That would mean the user removed the item added by kctrl
But yep makes sense I guess, will do these split up changes tomorrow

File: cli/pkg/kctrl/cmd/package/init.go (hunk 20 above)
Comment (praveenrewar) [reply cli/pkg/kctrl/cmd/package/init.go]: I am not sure how that will result in an exception. What I am suggesting is that instead of returning the sourceConfiguration, we directly call the sourceConfiguration.Configure in the main configure function based on the fetch option selected and then return the final result (an error).

File: cli/pkg/kctrl/local/sources/git.go (hunk 25 above)
Comment (praveenrewar) [reply cli/pkg/kctrl/local/sources/git.go]: Maybe we can do the `contents := g.vendirConfig...` in the functions where we are using `vendirConfig.Contents()[0]` more frequently?

File: cli/pkg/kctrl/cmd/package/init.go (hunk 20 above)
Comment (100mik) [reply cli/pkg/kctrl/cmd/package/init.go]: Oh I thought the you meant the function should handle the nil check we are doing

That is what we were doing before, It will also eliminate the need of having an interface, happy to do that 
Will make that change tomorrow

Comment (100mik) [reply cli/pkg/kctrl/cmd/package/init.go]: Done!

File: cli/pkg/kctrl/local/sources/vendir_runner.go (hunk 19 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/vendir_runner.go]: Looks like it does the exact same thing 😆 
Making the change

Comment (100mik) [reply cli/pkg/kctrl/local/sources/vendir_runner.go]: I do not see the point of using the exec.Result object either, it just adds `(see status.usefulErrormessage for details)` which we do not want

Comment (100mik) [reply cli/pkg/kctrl/local/sources/vendir_runner.go]: Done!

File: cli/pkg/kctrl/cmd/package/init.go (hunk 21 above)
Comment (100mik) [reply cli/pkg/kctrl/cmd/package/init.go]: Made this inline

File: cli/pkg/kctrl/cmd/package/init.go (hunk 22 above)
Comment (100mik) [reply cli/pkg/kctrl/cmd/package/init.go]: Same resolution as above

File: cli/pkg/kctrl/local/sources/git.go (hunk 24 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/git.go:57]: Looks like we are doing he same thing?
That is what `Contents` returns :/

File: cli/pkg/kctrl/local/sources/fetch.go (hunk 5 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/fetch.go]: Another case for having a type is that the check `fetchMode == LocalDirectory` is in a bunch of places and we will need to add cases if we introduce more local workflows.
Having something like `IsLocal` which has this condition really helps out , hm.

File: cli/pkg/kctrl/cmd/package/init.go [hunk 26]
```diff
@@ -64,54 +64,55 @@ func (o *InitOptions) Run() error {
...
 	if err != nil {
 		return err
 	}
 
-	pkgBuild.Save()
+	appSpec := pkgBuild.GetAppSpec()
+	if appSpec.Deploy == nil {
+		appSpec.Deploy = []kcv1alpha1.AppDeploy{{Kapp: &kcv1alpha1.AppDeployKapp{}}}
+	}
+	pkgBuild.SetAppSpec(appSpec)
```
Comment (praveenrewar) [inline cli/pkg/kctrl/cmd/package/init.go]: Maybe all of this should be under on function? (It seems like we are just setting deploy value to if it's nil).

File: cli/pkg/kctrl/cmd/package/repository/release/release.go [hunk 27]
```diff
@@ -184,33 +183,33 @@ func (o *ReleaseOptions) Run() error {
...
-			},
-			ObjectMeta: metav1.ObjectMeta{
-				CreationTimestamp: metav1.NewTime(time.Now()),
-			},
-			Spec: build.PackageRepoBuildSpec{
-				Export: &build.PackageRepoBuildExport{
-					ImgpkgBundle: &build.PackageRepoBuildExportImgpkgBundle{},
+	_, err := os.Stat(pkgRepoBuildFilePath)
+	if err != nil {
+		if os.IsNotExist(err) {
```
Comment (praveenrewar) [inline cli/pkg/kctrl/cmd/package/repository/release/release.go]:
```go
if !os.IsNotExist(err) {
      return &build.PackageRepoBuild{}, err
}
... snip ...
```

File: cli/pkg/kctrl/local/buildconfigs/pkg_build.go [hunk 28]
```diff
@@ -73,11 +86,25 @@ func (b *PackageBuild) SetObjectMeta(metaObj *metav1.ObjectMeta) {
...
 	return
 }
+
+func (b *PackageBuild) HasHelmTemplate() bool {
+	appSpec := b.getAppSpec()
+	if appSpec == nil || appSpec.Template == nil {
+		return false
+	}
+	appTemplates := appSpec.Template
+	for _, appTemplate := range appTemplates {
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/buildconfigs/pkg_build.go]: inline?

File: cli/pkg/kctrl/local/sources/template.go [hunk 29]
```diff
@@ -1,50 +1,40 @@
...
 	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/apis/kappctrl/v1alpha1"
 )
 
-const (
-	UpstreamFolderName = "upstream"
-	StdIn              = "-"
-)
-
-type TemplateStep struct {
+type TemplateConfiguration struct {
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/template.go]: Should this just be `Template`? (`TemplateConfiguration.Configure` sounds a bit weird)

File: cli/pkg/kctrl/local/sources/vendir_runner.go [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		r.ui.PrintCmdExecutionOutput(scanner.Text())
+	}
+	return nil
+}
+
+// TODO vendir sync failure. Reproduce: In case of 429 from github, we dont show errors today.
+func (r VendirRunner) runVendirSync() error {
```
Comment (praveenrewar) [inline cli/pkg/kctrl/local/sources/vendir_runner.go]: `runVendirSync` -> `sync`?

File: cli/pkg/kctrl/cmd/package/init.go (hunk 26 above)
Comment (100mik) [reply cli/pkg/kctrl/cmd/package/init.go]: Done!

File: cli/pkg/kctrl/cmd/package/repository/release/release.go (hunk 27 above)
Comment (100mik) [reply cli/pkg/kctrl/cmd/package/repository/release/release.go]: Done!

File: cli/pkg/kctrl/local/buildconfigs/pkg_build.go (hunk 28 above)
Comment (100mik) [reply cli/pkg/kctrl/local/buildconfigs/pkg_build.go]: Done!

File: cli/pkg/kctrl/local/sources/template.go (hunk 29 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/template.go]: Makes sense! I wonder if there could be a better name though

File: cli/pkg/kctrl/local/sources/vendir_runner.go (hunk 30 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/vendir_runner.go]: Done!

File: cli/pkg/kctrl/local/sources/template.go (hunk 29 above)
Comment (100mik) [reply cli/pkg/kctrl/local/sources/template.go]: Changed it to template for now

**Review by praveenrewar — APPROVED** [review]
LGTM!
